### PR TITLE
[Bugfix][Server][WFS] Server wfs format field

### PR DESCRIPTION
**Pull request by rldhont**
## Description

In schema document (.xsd) provided by WFS DescribeFeatureType request:
* Replace `double` by `decimal`
* Use `int`, `unsignedInt`, `long` and `unsignedLong` for `QVariant::Int`, `QVariant::UInt`, `QVariant::LongLong` and `QVariant::ULongLong`
* Define double with 0 precision to `integer`

Define `encodeValueToText` method to correctly format field values to XML text node.

Use editor widget setup for `DateTime` and `Range` fields.

## Checklist

- [x] Commit messages are descriptive and explain the rationale for changes
- [x] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [x] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [ ] New unit tests have been added for core changes
- [x] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTE.md#contributing-to-qgis) before each commit

### CODE REVIEW

**Comment by rldhont** [comment]
@nyalldawson part of this PR is a bugfix! I am working to do it for 3.4 !

**Comment by nyalldawson** [comment]
Ok - I was going purely off the "feature" tag, which would make it ineligible for 3.4 now.

File: src/server/services/wfs/qgswfsdescribefeaturetype.cpp [hunk 1]
```diff
@@ -262,27 +265,86 @@ namespace QgsWfs
...
+        }
         else
+        {
           attElem.setAttribute( QStringLiteral( "type" ), QStringLiteral( "string" ) );
+        }
+
+        const QgsEditorWidgetSetup setup = field.editorWidgetSetup();
+        if ( setup.type() ==  QStringLiteral( "DateTime" ) )
+        {
+          QgsDateTimeFieldFormatter *fieldFormatter = new QgsDateTimeFieldFormatter();
```
Comment (pblottiere) [inline src/server/services/wfs/qgswfsdescribefeaturetype.cpp]: It seems that there's a memory leak here (`fieldFormatter` is never deleted).

BTW, a pointer is not necessary in this case.

File: src/server/services/wfs/qgswfsgetfeature.cpp [hunk 2]
```diff
@@ -1430,18 +1437,87 @@ namespace QgsWfs
...
     }
 
+    QString encodeValueToText( const QVariant &value, const QgsEditorWidgetSetup &setup )
+    {
+      if ( value.isNull() )
+        return QStringLiteral( "null" );
+
+      if ( setup.type() ==  QStringLiteral( "DateTime" ) )
+      {
+        QgsDateTimeFieldFormatter *fieldFormatter = new QgsDateTimeFieldFormatter();
```
Comment (pblottiere) [inline src/server/services/wfs/qgswfsgetfeature.cpp]: The same remark as above applies (memory leak / pointer?)

File: src/server/services/wfs/qgswfsgetfeature.cpp [hunk 3]
```diff
@@ -1430,18 +1437,87 @@ namespace QgsWfs
...
+
+        case QVariant::StringList:
+        case QVariant::List:
+        case QVariant::Map:
+        {
+          QString v = QgsJsonUtils::encodeValue( value );
+
+          //do we need CDATA
+          if ( v.indexOf( '<' ) != -1 || v.indexOf( '&' ) != -1 )
+            v.prepend( "<![CDATA[" ).append( "]]>" );
```
Comment (pblottiere) [inline src/server/services/wfs/qgswfsgetfeature.cpp]: QStringLiteral?

File: src/server/services/wfs/qgswfsgetfeature.cpp [hunk 4]
```diff
@@ -1430,18 +1437,87 @@ namespace QgsWfs
...
+        }
+
+        default:
+        case QVariant::String:
+        {
+          QString v = value.toString();
+
+          //do we need CDATA
+          if ( v.indexOf( '<' ) != -1 || v.indexOf( '&' ) != -1 )
+            v.prepend( "<![CDATA[" ).append( "]]>" );
```
Comment (pblottiere) [inline src/server/services/wfs/qgswfsgetfeature.cpp]: QStringLiteral?

**Comment by rldhont** [comment]
@pblottiere thanks for review, the code has been updated.

File: src/server/services/wfs/qgswfsdescribefeaturetype.cpp [hunk 5]
```diff
@@ -262,27 +265,86 @@ namespace QgsWfs
...
           attElem.setAttribute( QStringLiteral( "type" ), QStringLiteral( "string" ) );
+        }
+
+        const QgsEditorWidgetSetup setup = field.editorWidgetSetup();
+        if ( setup.type() ==  QStringLiteral( "DateTime" ) )
+        {
+          QgsDateTimeFieldFormatter fieldFormatter;
+          const QVariantMap config = setup.config();
+          const QString fieldFormat = config.value( QStringLiteral( "field_format" ), fieldFormatter.defaultFormat( field.type() ) ).toString();
+          if ( fieldFormat == "yyyy-MM-dd" )
```
Comment (pblottiere) [inline src/server/services/wfs/qgswfsdescribefeaturetype.cpp]: QStringLiteral (and some others below)

**Comment by pblottiere** [comment]
Code looks good, thanks @rldhont!

**Comment by pblottiere** [comment]
@rldhont The label  `Requires Tests!` was set on this PR, but I don't see any...

**Comment by rldhont** [comment]
Yes I need to check the available layers for this ;-)

**Pull request merged**